packages/api/amplify_api/lib/src/graphql/ws/websocket_message.dart

```diff
@@ -0,0 +1,202 @@
+    final connectionTimeoutMs = json['connectionTimeoutMs'] as int;
+    return ConnectionAckMessagePayload(connectionTimeoutMs);
+  }
+
+  @override
+  Map<String, dynamic> toJson() => <String, dynamic>{
+        'connectionTimeoutMs': connectionTimeoutMs,
+      };
+}
+
+class SubscriptionRegistrationPayload extends WebSocketMessagePayload {
+  final GraphQLRequest request;
+  final AWSApiConfig config;
+
+  SubscriptionRegistrationPayload({
+    required this.request,
+    required this.config,
+  });
+
+  @override
+  Map<String, dynamic> toJson() {
+    return <String, dynamic>{
+      'data': jsonEncode(
+          {'variables': request.variables, 'query': request.document}),
+      'extensions': <String, dynamic>{'authorization': authorizationHeaders}
+    };
+  }
+}
+
+class SubscriptionDataPayload extends WebSocketMessagePayload {
+  final Map<String, dynamic>? data;
+  final Map<String, dynamic>? errors;
+
+  SubscriptionDataPayload(this.data, this.errors);
+
+  static SubscriptionDataPayload fromJson(Map json) {
+    final data = json['data'] as Map?;
+    final errors = json['errors'] as Map?;
+    return SubscriptionDataPayload(
+      data?.cast<String, dynamic>(),
+      errors?.cast<String, dynamic>(),
+    );
+  }
+
+  @override
+  Map<String, dynamic> toJson() => <String, dynamic>{
+        'data': data,
+        'errors': errors,
+      };
+}
+
+class WebSocketError extends WebSocketMessagePayload implements Exception {
+  final List<Map> errors;
+
+  WebSocketError(this.errors);
+
+  static WebSocketError fromJson(Map json) {
+    final errors = json['errors'] as List?;
+    return WebSocketError(errors?.cast() ?? []);
+  }
+
+  @override
+  Map<String, dynamic> toJson() => <String, dynamic>{
+        'errors': errors,
+      };
+}
+
+class WebSocketMessage {
+  final String? id;
+  final MessageType messageType;
+  final WebSocketMessagePayload? payload;
+
+  WebSocketMessage({
+    String? id,
+    required this.messageType,
+    this.payload,
+  }) : id = id ?? UUID.getUUID();
+
+  WebSocketMessage._({
+    this.id,
+    required this.messageType,
+    this.payload,
+  });
+
+  static WebSocketMessage fromJson(Map json) {
+    final id = json['id'] as String?;
+    final type = json['type'] as String;
+    final messageType = MessageType.fromJson(type);
+    final payloadMap = json['payload'] as Map?;
+    final payload = payloadMap == null
+        ? null
+        : WebSocketMessagePayload.fromJson(
+            payloadMap,
+            messageType,
+          );
+    return WebSocketMessage._(
+      id: id,
+      messageType: messageType,
+      payload: payload,
+    );
+  }
+
+  Map<String, dynamic> toJson() => <String, dynamic>{
+        if (id != null) 'id': id,
+        'type': messageType.type,
+        if (payload != null) 'payload': payload?.toJson(),
+      };
+
+  @override
+  String toString() {
+    return prettyPrintJson(this);
+  }
+}
+
+class WebSocketConnectionInitMessage extends WebSocketMessage {
```

Comment:
Following the logic on these classes is really difficult since they get mutated after instantiation. I would suggest making them all `const`. This will force a better data flow and more functional programming.

Reply:
I think that's a good suggestion. I haven't made them const yet bc they have a uuid that gets generated if not passed by caller, but I changed the logic to avoid mutating them and generally refactored the auth logic.

Reply:
Yeah, if not `const` then you can mark them `@immutable` which will help catch the same issues.